src/module.c

```diff
@@ -7388,6 +7405,262 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+ * If none of the callbacks authenticate or deny auth, then password based auth is attemmpted
+ * and will authenticate or add failure logs and reply to the clients accordingly. */
+void RM_RegisterCustomAuthCallback(RedisModuleCtx *ctx, RedisModuleCustomAuthCallback cb) {
+    RedisModuleCustomAuthCtx *auth_ctx = zmalloc(sizeof(RedisModuleCustomAuthCtx));
+    auth_ctx->module = ctx->module;
+    auth_ctx->auth_cb = cb;
+    listAddNodeTail(moduleCustomAuthCallbacks, auth_ctx);
+}
+
+/* Returns 1 if the module has any custom auth cbs registered. If `should_unregister` is  non zero,
+ * all the module's custom auth cbs are unregistered.
+ * Returns 0 otherwise - if the Module has no custom auth cb registered. */
+int moduleSearchOrUnregisterCustomAuthCBs(RedisModule *module, int should_unregister) {
+    int callback_found = 0;
+    listIter li;
+    listNode *ln;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    while ((ln = listNext(&li))) {
+        RedisModuleCustomAuthCtx *ctx = listNodeValue(ln);
+        if (ctx->module == module) {
+            callback_found = 1;
+            if (!should_unregister) return callback_found;
+            listDelNode(moduleCustomAuthCallbacks, ln);
+            zfree(ctx);
+        }
+    }
+    return callback_found;
+}
+
+/* Returns 1 if custom module based auth is in progress on any client. Returns 0 otherwise. */
+int isCustomAuthInProgress() {
+    if (!listLength(moduleCustomAuthCallbacks)) return 0;
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (c->flags & CLIENT_CUSTOM_AUTH) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* Unregister ALL the custom auth callbacks that this module had registered.
+ * Returns REDISMODULE_ERR in the following cases:
+ * - `ctx` is NULL.
+ * - ENOENT: The module does not have any custom auth cbs registered.
+ * - EINPROGRESS: Custom Auth is in progress and the module has a custom auth cb registered.
+ * Returns REDISMODULE_OK if the module had custom auth cbs and they were unregistered. */
+int RM_UnregisterCustomAuthCallbacks(RedisModuleCtx *ctx) {
+    if (!ctx) {
+        return REDISMODULE_ERR;
+    }
+    /* Search and return early if the Module does not have any custom auth cbs registered. */
+    if (!moduleSearchOrUnregisterCustomAuthCBs(ctx->module, 0)) {
+        errno = ENOENT;
+        return REDISMODULE_ERR;
+    }
+    if (isCustomAuthInProgress()) {
+        errno = EINPROGRESS;
+        return REDISMODULE_ERR;
+    }
+    moduleSearchOrUnregisterCustomAuthCBs(ctx->module, 1);
+    return REDISMODULE_OK;
+}
+
+/* Returns 0 when custom auth is in progress with a blocking implementation OR when it has been
+ * concluded either with success or explicitly denied.
+ * Returns 1 otherwise - when we should attempt the next custom auth callback if one exists. */
+int tryNextCustomAuthCB(client *c, int cb_result) {
```

Comment:
This seems like a highly coupled function, I would have just left this in the calling function site. You also left a comment at the calling site describing the functionality of this function, which isn't great.

Reply:
This function is removed in https://github.com/redis/redis/pull/11659/commits/b21ad243127fddee6889652e579a2d15fa725aed